crates/agent/src/publications/specs.rs

```diff
@@ -350,6 +350,71 @@ pub fn validate_transition(
     errors
 }
 
+pub async fn enforce_resource_quotas(
+    spec_rows: &[SpecRow],
+    txn: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+) -> anyhow::Result<Vec<Error>> {
+    let tenants = spec_rows
+        .iter()
+        .map(|spec_row| tenant_name_for_catalog_name(spec_row.catalog_name.clone()))
+        .collect::<Result<HashSet<_>, _>>()?;
+
+    let errors = agent_sql::publications::find_tenant_quotas(tenants.into_iter().collect(), txn)
+        .await?
+        .into_iter()
+        .flat_map(|tenant| {
+            let mut errs = vec![];
+            if tenant.collections_used > tenant.collections_quota {
+                errs.push(format!(
+                    "Tenant {} would exceed collection quota of {} by {}",
+                    tenant.name,
+                    tenant.collections_quota,
+                    (tenant.collections_used - tenant.collections_quota)
+                ))
+            }
+            if tenant.derivations_used > tenant.derivations_quota {
+                errs.push(format!(
+                    "Tenant {} would exceed derivation quota of {} by {}",
+                    tenant.name,
+                    tenant.derivations_quota,
+                    (tenant.derivations_used - tenant.derivations_quota)
+                ))
+            }
+            if tenant.captures_used > tenant.captures_quota {
+                errs.push(format!(
+                    "Tenant {} would exceed capture quota of {} by {}",
+                    tenant.name,
+                    tenant.captures_quota,
+                    (tenant.captures_used - tenant.captures_quota)
+                ))
+            }
+            if tenant.materializations_used > tenant.materializations_quota {
+                errs.push(format!(
+                    "Tenant {} would exceed materialization quota of {} by {}",
+                    tenant.name,
+                    tenant.materializations_quota,
+                    (tenant.materializations_used - tenant.materializations_quota)
+                ))
+            }
+            errs
+        })
+        .map(|err_str| Error {
+            detail: err_str,
+            ..Default::default()
+        })
+        .collect();
+
+    Ok(errors)
+}
+
+pub fn tenant_name_for_catalog_name(catalog_name: String) -> anyhow::Result<String> {
+    catalog_name
```

Comment:
There's a sneaky bug here, which is that this split will drop the trailing '/' and thus won't match any entries of the `tenants` table, which always have a trailing `/`.

An optional suggestion: let the database do more of the heavy lift here. Pass in the spec IDs as `resolve_expanded_rows` does and use a CTE as part of `find_tenant_quotas` which maps that into unique tenants. I think you'll find that this construction is easier to test, and it also removes the possibility of mismatches like this.

Reply:
Done!